keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md

```diff
@@ -125,29 +124,62 @@ serves the resourceVersion="0" list requests from reflectors today.
 
-#### Use RequestProgress to enable automatic watch updates
+A single request is not enough to ensure that the wait on resource version terminates.
+If etcd receives a burst of events, the watch might be in the middle of processing a large batch of events when the watch cache requests progress.
+The resulting progress notification will return a resource version earlier than requested, which could cause an infinite wait if there are no follow-up events.
+
+To rely on progress notification for freshness, the watch cache will periodically (every 100ms) request progress until it reaches the requested revision.
+
+#### The algorithm
 
 When a consistent LIST request is received and the watch cache is enabled:
 
 - Get the current revision from etcd for the resource type being served.
-  Use the [getCurrentResourceVersionFromStorage] added as part of [Watch-List KEP].
 - If the cache already has the current revision, serve the request from cache. If not,
-  - Send a `WatchProgressRequest` to etcd on the watch channel that the watch cache is consuming.
-- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly for the watch to catch up to the current revision.
+  - Add a "waiting read". That should trigger a background goroutine that will
+    - Send `WatchProgressRequest` request to etcd every 100ms as long there is at least one waiting read.
+  - Wait for the watch to catch up to the revision.
+  - Remove a "waiting read".
 - If the block times out, the request will result in rejection. (see "What if the watch cache is stale?" section for details)
 
 Consistent GET requests will continue to be served directly from etcd. We will
 only serve consistent LIST requests from cache.
 
-[getCurrentResourceVersionFromStorage]: https://github.com/kubernetes/kubernetes/blob/3f247e59edfd4083242ad7271d076a38291760ff/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1246-L1278
 [Watch-List KEP]: /keps/sig-api-machinery/3157-watch-list
 
+### Bug in etcd progress notification
+
+Only recently community discovered a bug [etcd-io/etcd#15220] for requesting progress notification.
+This bug causes a race between sending an event and progress notification with the same revision.
+Normally we would expect to event to always come first, however due to the bug, progress notification might be sent first.
+Serving a consistent lists in that situation could result in returning too early and missing the event.
+The bug was only fixed in v3.4.25 and v3.5.8 (4 months old), which is too fresh for Kubernetes to not handle.
+
+We propose to implement a direct mitigation for this bug. 
+Watch cache will not trust the last revision send by progress notification.
+When it receives a progress notification, watch cache would reduce the received resourceVersion by one, and save it instead of the original one.
+To prevent infinite wait on resources with no change, we will trust a second progress notification with a patching resource version.
```

Comment:
So effectively the check is:
```
if wc.resourceVersion < progressNotifyRV {
  wc.resourceVersion = progressNotifyRV - 1
} else {
  wc.resourceVersion = max(wc.resourceVersion, progressNotify)
```

right?

It potentially adds 100ms additional latency, but I think it's acceptable...

The only question I have is - why do we have guarantee that the second notification cannot suffer from the exact same problem?

Reply:
I still don't fully understand why the second notification cannot suffer from the same problem...

IIUC, the problem is a race [between sending the event and the progress-notify].

So sending another one using the same mechanism 100ms later hugely reduces the probability, but in theory in super heavily loaded environment doesn't reduce it down to zero, no?

Reply:
Our current assumption is that etcd race is just between event and single progress notification sharing the same revision. Question is what happens if there are multiple progress notification requested in quick succession and whether it can happen in Kubernetes.  

Will do some experiments and let you know.

Reply:
So where we are with it?

Reply:
Run robustness tests with progress notification requested every 10 Microseconds (lower caused a deadlock, didn't investigate why). Results:
```
    137 progress_notify, rev: 1
     67 progress_notify, rev: 2
    144 progress_notify, rev: 3
      3 progress_notify, rev: 4
      2 progress_notify, rev: 5
      1 progress_notify, rev: 6
     12 progress_notify, rev: 7
     29 progress_notify, rev: 8
     54 progress_notify, rev: 9
    229 progress_notify, rev: 10
    153 progress_notify, rev: 11
     61 progress_notify, rev: 12
      2 progress_notify, rev: 13
     63 progress_notify, rev: 14
      9 progress_notify, rev: 15
      8 progress_notify, rev: 16
     30 progress_notify, rev: 17
     62 progress_notify, rev: 18
      6 progress_notify, rev: 19
      1 progress_notify, rev: 20
     87 progress_notify, rev: 21
     56 progress_notify, rev: 22
      1 progress_notify, rev: 23
      4 progress_notify, rev: 24
      1 progress_notify, rev: 25
      1 progress_notify, rev: 26
     77 progress_notify, rev: 27
     82 progress_notify, rev: 28
      4 progress_notify, rev: 29
      9 progress_notify, rev: 30
     46 progress_notify, rev: 31
     47 progress_notify, rev: 32
     60 progress_notify, rev: 33
      1 progress_notify, rev: 34
     80 progress_notify, rev: 35
      1 progress_notify, rev: 36
     61 progress_notify, rev: 37
     69 progress_notify, rev: 38
    241 progress_notify, rev: 39
     77 progress_notify, rev: 40
     13 progress_notify, rev: 41
     62 progress_notify, rev: 42
      4 progress_notify, rev: 43
     70 progress_notify, rev: 44
     66 progress_notify, rev: 45
      5 progress_notify, rev: 46
      4 progress_notify, rev: 47
     61 progress_notify, rev: 48
     70 progress_notify, rev: 49
      9 progress_notify, rev: 50
     76 progress_notify, rev: 51
     12 progress_notify, rev: 52
     60 progress_notify, rev: 53
      1 progress_notify, rev: 54
      4 progress_notify, rev: 55
     70 progress_notify, rev: 56
      5 progress_notify, rev: 57
      1 progress_notify, rev: 58
     68 progress_notify, rev: 59
     73 progress_notify, rev: 60
     21 progress_notify, rev: 61
     38 progress_notify, rev: 62
    421 progress_notify, rev: 63
      3 progress_notify, rev: 64
      4 progress_notify, rev: 65
      4 progress_notify, rev: 66
     41 progress_notify, rev: 67
     57 progress_notify, rev: 68
    113 progress_notify, rev: 69
      1 event, rev: 2
      1 event, rev: 3
      1 event, rev: 4
      1 event, rev: 5
      1 event, rev: 6
      1 event, rev: 7
      1 event, rev: 8
      1 event, rev: 9
      1 event, rev: 10
      1 event, rev: 11
      1 event, rev: 12
      1 event, rev: 13
      1 event, rev: 14
      1 event, rev: 15
      1 event, rev: 16
      1 event, rev: 17
      1 event, rev: 18
      1 event, rev: 19
      1 event, rev: 20
      1 event, rev: 21
      1 event, rev: 22
      1 event, rev: 23
      1 event, rev: 24
      1 event, rev: 25
      1 event, rev: 26
      1 event, rev: 27
      1 event, rev: 28
      2 event, rev: 29
      1 event, rev: 30
      1 event, rev: 31
      1 event, rev: 32
      1 event, rev: 33
      1 event, rev: 34
      4 event, rev: 35
      1 event, rev: 36
      1 event, rev: 37
      1 event, rev: 38
      1 event, rev: 39
      3 event, rev: 40
      1 event, rev: 41
      2 event, rev: 42
      1 event, rev: 43
      1 event, rev: 44
      3 event, rev: 45
      1 event, rev: 46
      1 event, rev: 47
      1 event, rev: 48
      1 event, rev: 49
      1 event, rev: 50
      1 event, rev: 51
      1 event, rev: 52
      1 event, rev: 53
      1 event, rev: 54
      1 event, rev: 55
      1 event, rev: 56
      1 event, rev: 57
      1 event, rev: 58
      1 event, rev: 59
      2 event, rev: 60
      1 event, rev: 61
      1 event, rev: 62
      1 event, rev: 63
      1 event, rev: 64
      1 event, rev: 65
      1 event, rev: 66
      1 event, rev: 67
      1 event, rev: 68
      3 event, rev: 69
    123 progress_notify, rev: 69
      1 event, rev: 70
      5 progress_notify, rev: 70
      1 event, rev: 71
      1 event, rev: 72
     96 progress_notify, rev: 72
      1 event, rev: 73
    140 progress_notify, rev: 73
      1 event, rev: 74
      6 progress_notify, rev: 74
      1 event, rev: 75
     11 progress_notify, rev: 75
      1 event, rev: 76
      1 progress_notify, rev: 75
      3 progress_notify, rev: 76
      1 event, rev: 77
     76 progress_notify, rev: 77
      1 event, rev: 78
    256 progress_notify, rev: 78
      1 event, rev: 79
      4 progress_notify, rev: 79
      1 event, rev: 80
      2 progress_notify, rev: 79
      1 progress_notify, rev: 80
      1 event, rev: 81
      1 progress_notify, rev: 80
      7 progress_notify, rev: 81
      1 event, rev: 82
     64 progress_notify, rev: 82
      3 event, rev: 83
    267 progress_notify, rev: 83
      1 event, rev: 84
      4 progress_notify, rev: 84
      1 event, rev: 85
     19 progress_notify, rev: 85
      1 event, rev: 86
      1 progress_notify, rev: 85
     70 progress_notify, rev: 86
      4 event, rev: 87
      5 progress_notify, rev: 87
      1 event, rev: 88
      5 progress_notify, rev: 88
      1 event, rev: 89
     83 progress_notify, rev: 89
      2 event, rev: 90
     66 progress_notify, rev: 90
      1 event, rev: 91
      2 progress_notify, rev: 90
      3 progress_notify, rev: 91
      1 event, rev: 92
      1 event, rev: 93
      1 progress_notify, rev: 92
     66 progress_notify, rev: 93
      1 event, rev: 94
      2 progress_notify, rev: 94
      1 event, rev: 95
     64 progress_notify, rev: 95
      1 event, rev: 96
     65 progress_notify, rev: 96
      1 event, rev: 97
      4 progress_notify, rev: 97
      1 event, rev: 98
    246 progress_notify, rev: 98
      1 event, rev: 99
      1 progress_notify, rev: 99
      1 event, rev: 100
      1 progress_notify, rev: 99
      4 progress_notify, rev: 100
      1 event, rev: 101
     67 progress_notify, rev: 101
      1 event, rev: 102
      2 progress_notify, rev: 102
      1 event, rev: 103
     61 progress_notify, rev: 103
      1 progress_notify, rev: 104
      1 event, rev: 104
    219 progress_notify, rev: 104
      1 event, rev: 105
    388 progress_notify, rev: 105
      1 event, rev: 106
    380 progress_notify, rev: 106
      1 event, rev: 107
    856 progress_notify, rev: 107
      1 event, rev: 108
    143 progress_notify, rev: 108
      1 event, rev: 109
    577 progress_notify, rev: 109
      1 event, rev: 110
    195 progress_notify, rev: 110
      1 event, rev: 111
    338 progress_notify, rev: 111
      1 event, rev: 112
    182 progress_notify, rev: 112
      1 event, rev: 113
    582 progress_notify, rev: 113
      2 event, rev: 114
    687 progress_notify, rev: 114
      1 event, rev: 115
    364 progress_notify, rev: 115
      1 event, rev: 116
    192 progress_notify, rev: 116
      1 event, rev: 117
    772 progress_notify, rev: 117
      1 event, rev: 118
   1305 progress_notify, rev: 118
      1 event, rev: 119
      1 progress_notify, rev: 118
    564 progress_notify, rev: 119
      1 event, rev: 120
      1 progress_notify, rev: 119
    186 progress_notify, rev: 120
      3 event, rev: 121
    381 progress_notify, rev: 121
      1 event, rev: 122
    202 progress_notify, rev: 122
      1 event, rev: 123
    158 progress_notify, rev: 123
      1 event, rev: 124
    553 progress_notify, rev: 124
      1 event, rev: 125
     70 progress_notify, rev: 125
      2 event, rev: 126
   1053 progress_notify, rev: 126
      1 event, rev: 127
      3 progress_notify, rev: 126
    201 progress_notify, rev: 127
      1 event, rev: 128
      1 progress_notify, rev: 127
    376 progress_notify, rev: 128
      1 event, rev: 129
    558 progress_notify, rev: 129
      1 event, rev: 130
   1061 progress_notify, rev: 130
      1 event, rev: 131
    728 progress_notify, rev: 131
      1 event, rev: 132
      1 progress_notify, rev: 131
    593 progress_notify, rev: 132
      1 event, rev: 133
    216 progress_notify, rev: 133
      1 event, rev: 134
    207 progress_notify, rev: 134
      1 event, rev: 135
    195 progress_notify, rev: 135
      1 event, rev: 136
    561 progress_notify, rev: 136
      1 event, rev: 137
    375 progress_notify, rev: 137
      1 event, rev: 138
      1 progress_notify, rev: 137
    382 progress_notify, rev: 138
      1 event, rev: 139
    332 progress_notify, rev: 139
      1 event, rev: 140
    162 progress_notify, rev: 140
      1 event, rev: 141
    597 progress_notify, rev: 141
      1 event, rev: 142
      1 progress_notify, rev: 141
    613 progress_notify, rev: 142
      2 event, rev: 143
      1 progress_notify, rev: 142
    729 progress_notify, rev: 143
      1 event, rev: 144
    158 progress_notify, rev: 144
      1 event, rev: 145
   1061 progress_notify, rev: 145
      1 event, rev: 146
    175 progress_notify, rev: 146
      1 event, rev: 147
      1 progress_notify, rev: 146
    408 progress_notify, rev: 147
      2 event, rev: 148
    545 progress_notify, rev: 148
      1 event, rev: 149
      1 progress_notify, rev: 148
    191 progress_notify, rev: 149
      1 event, rev: 150
    250 progress_notify, rev: 150
      1 event, rev: 151
     10 progress_notify, rev: 150
    282 progress_notify, rev: 151
      1 event, rev: 152
    190 progress_notify, rev: 152
      3 event, rev: 153
    791 progress_notify, rev: 153
      1 event, rev: 154
    172 progress_notify, rev: 154
      1 event, rev: 155
    194 progress_notify, rev: 155
      1 event, rev: 156
    909 progress_notify, rev: 156
      1 event, rev: 157
    354 progress_notify, rev: 157
      2 event, rev: 158
    187 progress_notify, rev: 158
      1 event, rev: 159
    399 progress_notify, rev: 159
      1 event, rev: 160
    169 progress_notify, rev: 160
      1 event, rev: 161
    177 progress_notify, rev: 161
      1 event, rev: 162
    718 progress_notify, rev: 162
      1 event, rev: 163
    230 progress_notify, rev: 163
      1 event, rev: 164
    713 progress_notify, rev: 164
      1 event, rev: 165
    548 progress_notify, rev: 165
      1 event, rev: 166
      2 progress_notify, rev: 165
   1301 progress_notify, rev: 166
      1 event, rev: 167
    361 progress_notify, rev: 167
      1 event, rev: 168
    192 progress_notify, rev: 168
      1 event, rev: 169
      1 progress_notify, rev: 168
     50 progress_notify, rev: 169
      1 event, rev: 170
      9 progress_notify, rev: 169
     47 progress_notify, rev: 170
      1 event, rev: 171
     12 progress_notify, rev: 170
    497 progress_notify, rev: 171
      1 event, rev: 172
    178 progress_notify, rev: 172
      1 event, rev: 173
    177 progress_notify, rev: 173
      1 event, rev: 174
    179 progress_notify, rev: 174
      3 event, rev: 175
    211 progress_notify, rev: 175
      1 event, rev: 176
      4 progress_notify, rev: 175
   1108 progress_notify, rev: 176
      1 event, rev: 177
    173 progress_notify, rev: 177
      1 event, rev: 178
     28 progress_notify, rev: 178
      1 event, rev: 179
     17 progress_notify, rev: 178
      2 progress_notify, rev: 179
      2 progress_notify, rev: 180
      1 event, rev: 180
     17 progress_notify, rev: 180
      1 event, rev: 181
      4 progress_notify, rev: 180
     38 progress_notify, rev: 181
      1 event, rev: 182
      1 progress_notify, rev: 181
     12 progress_notify, rev: 182
      1 event, rev: 183
      1 progress_notify, rev: 182
    158 progress_notify, rev: 183
      1 event, rev: 184
     16 progress_notify, rev: 183
    118 progress_notify, rev: 184
      1 event, rev: 185
      8 progress_notify, rev: 185
      1 event, rev: 186
     81 progress_notify, rev: 186
      1 event, rev: 187
     71 progress_notify, rev: 187
      1 event, rev: 188
      1 progress_notify, rev: 187
     81 progress_notify, rev: 188
      1 event, rev: 189
     70 progress_notify, rev: 189
      1 event, rev: 190
      1 progress_notify, rev: 189
      1 event, rev: 191
      1 progress_notify, rev: 190
     11 progress_notify, rev: 191
      1 event, rev: 192
      1 progress_notify, rev: 192
      1 event, rev: 193
      1 progress_notify, rev: 192
      1 progress_notify, rev: 193
      1 event, rev: 194
     61 progress_notify, rev: 194
      1 event, rev: 195
     64 progress_notify, rev: 195
      1 event, rev: 196
      7 progress_notify, rev: 196
      4 event, rev: 197
      4 progress_notify, rev: 197
      1 event, rev: 198
    108 progress_notify, rev: 198
      1 event, rev: 199
     56 progress_notify, rev: 199
      1 event, rev: 200
      8 progress_notify, rev: 200
      1 event, rev: 201
     63 progress_notify, rev: 201
      1 event, rev: 202
    258 progress_notify, rev: 202
      1 event, rev: 203
      4 progress_notify, rev: 203
      1 event, rev: 204
      6 progress_notify, rev: 204
      1 event, rev: 205
     77 progress_notify, rev: 205
      1 event, rev: 206
     63 progress_notify, rev: 206
      1 event, rev: 207
      3 progress_notify, rev: 207
      1 event, rev: 208
     96 progress_notify, rev: 208
      1 event, rev: 209
    242 progress_notify, rev: 209
      1 event, rev: 210
      3 progress_notify, rev: 210
      3 event, rev: 211
      5 progress_notify, rev: 210
      3 progress_notify, rev: 211
      1 event, rev: 212
      2 progress_notify, rev: 211
     61 progress_notify, rev: 212
      1 event, rev: 213
     64 progress_notify, rev: 213
      1 event, rev: 214
      1 event, rev: 215
     17 progress_notify, rev: 215
      1 event, rev: 216
      2 progress_notify, rev: 215
     10 progress_notify, rev: 216
      1 event, rev: 217
     67 progress_notify, rev: 217
      1 event, rev: 218
     62 progress_notify, rev: 218
      2 event, rev: 219
      6 progress_notify, rev: 219
      1 event, rev: 220
      3 progress_notify, rev: 220
      1 event, rev: 221
     69 progress_notify, rev: 221
      1 event, rev: 222
     66 progress_notify, rev: 222
      1 event, rev: 223
      6 progress_notify, rev: 223
      1 event, rev: 224
      6 progress_notify, rev: 224
      1 event, rev: 225
      1 progress_notify, rev: 224
      1 event, rev: 226
      1 progress_notify, rev: 225
     67 progress_notify, rev: 226
      1 event, rev: 227
    226 progress_notify, rev: 227
      1 event, rev: 228
      1 progress_notify, rev: 227
     69 progress_notify, rev: 228
      2 event, rev: 229
      2 progress_notify, rev: 229
      1 event, rev: 230
      8 progress_notify, rev: 230
      1 event, rev: 231
    125 progress_notify, rev: 231
      1 event, rev: 232
      1 event, rev: 233
      1 progress_notify, rev: 232
      1 progress_notify, rev: 233
      1 event, rev: 234
      1 progress_notify, rev: 233
      1 event, rev: 235
      1 progress_notify, rev: 234
     58 progress_notify, rev: 235
      1 event, rev: 236
      3 progress_notify, rev: 235
     61 progress_notify, rev: 236
      1 event, rev: 237
      4 progress_notify, rev: 237
      1 event, rev: 238
     73 progress_notify, rev: 238
      1 event, rev: 239
      2 progress_notify, rev: 239
      1 event, rev: 240
     73 progress_notify, rev: 240
      1 event, rev: 241
     74 progress_notify, rev: 241
      1 event, rev: 242
      5 progress_notify, rev: 242
      1 event, rev: 243
     67 progress_notify, rev: 243
      2 progress_notify, rev: 244
      1 event, rev: 244
      2 progress_notify, rev: 244
      2 event, rev: 245
     65 progress_notify, rev: 245
      2 event, rev: 246
     68 progress_notify, rev: 246
      3 event, rev: 247
      3 progress_notify, rev: 247
      1 event, rev: 248
    233 progress_notify, rev: 248
      1 event, rev: 249
     79 progress_notify, rev: 249
      1 event, rev: 250
     21 progress_notify, rev: 250
      1 event, rev: 251
      3 progress_notify, rev: 251
      2 event, rev: 252
      3 progress_notify, rev: 252
      1 event, rev: 253
    174 progress_notify, rev: 253
      1 event, rev: 254
     12 progress_notify, rev: 253
      7 progress_notify, rev: 254
      1 event, rev: 255
     11 progress_notify, rev: 254
     19 progress_notify, rev: 255
      1 event, rev: 256
      7 progress_notify, rev: 255
    284 progress_notify, rev: 256
      4 event, rev: 257
     56 progress_notify, rev: 257
      1 event, rev: 258
      4 progress_notify, rev: 258
      3 event, rev: 259
     68 progress_notify, rev: 259
      1 event, rev: 260
     72 progress_notify, rev: 260
      1 event, rev: 261
      1 progress_notify, rev: 260
      7 progress_notify, rev: 261
      1 event, rev: 262
     67 progress_notify, rev: 262
      2 event, rev: 263
     66 progress_notify, rev: 263
      1 progress_notify, rev: 264
      1 event, rev: 264
     67 progress_notify, rev: 264
      1 event, rev: 265
     63 progress_notify, rev: 265
      1 event, rev: 266
      1 event, rev: 267
      1 progress_notify, rev: 266
      5 progress_notify, rev: 267
      1 event, rev: 268
     63 progress_notify, rev: 268
      1 event, rev: 269
    952 progress_notify, rev: 269
      1 event, rev: 270
    561 progress_notify, rev: 270
      1 event, rev: 271
    366 progress_notify, rev: 271
      1 event, rev: 272
    176 progress_notify, rev: 272
      1 event, rev: 273
    188 progress_notify, rev: 273
      1 event, rev: 274
      2 progress_notify, rev: 273
    349 progress_notify, rev: 274
      1 event, rev: 275
    160 progress_notify, rev: 275
      1 event, rev: 276
      1 progress_notify, rev: 275
    224 progress_notify, rev: 276
      1 event, rev: 277
     12 progress_notify, rev: 276
   1700 progress_notify, rev: 277
      1 event, rev: 278
    604 progress_notify, rev: 278
      1 event, rev: 279
    644 progress_notify, rev: 279
      1 event, rev: 280
    385 progress_notify, rev: 280
      1 event, rev: 281
      1 progress_notify, rev: 280
    267 progress_notify, rev: 281
      1 event, rev: 282
      3 progress_notify, rev: 281
   1864 progress_notify, rev: 282
      1 event, rev: 283
    920 progress_notify, rev: 283
      1 event, rev: 284
    181 progress_notify, rev: 284
      1 event, rev: 285
   1318 progress_notify, rev: 285
      1 event, rev: 286
      6 progress_notify, rev: 285
    641 progress_notify, rev: 286
      1 event, rev: 287
    361 progress_notify, rev: 287
      1 event, rev: 288
    567 progress_notify, rev: 288
      3 event, rev: 289
     73 progress_notify, rev: 289
      1 event, rev: 290
    336 progress_notify, rev: 290
      1 event, rev: 291
    363 progress_notify, rev: 291
      1 event, rev: 292
      1 progress_notify, rev: 291
    784 progress_notify, rev: 292
      1 event, rev: 293
    362 progress_notify, rev: 293
      1 event, rev: 294
    192 progress_notify, rev: 294
      1 event, rev: 295
      1 progress_notify, rev: 294
    249 progress_notify, rev: 295
      1 event, rev: 296
      9 progress_notify, rev: 295
   1120 progress_notify, rev: 296
      3 event, rev: 297
    185 progress_notify, rev: 297
      1 event, rev: 298
   1261 progress_notify, rev: 298
      1 event, rev: 299
    376 progress_notify, rev: 299
      1 event, rev: 300
    174 progress_notify, rev: 300
      1 event, rev: 301
      1 progress_notify, rev: 300
    181 progress_notify, rev: 301
      1 event, rev: 302
    371 progress_notify, rev: 302
      1 event, rev: 303
    567 progress_notify, rev: 303
      1 event, rev: 304
    195 progress_notify, rev: 304
      1 event, rev: 305
   1363 progress_notify, rev: 305
      1 event, rev: 306
   1151 progress_notify, rev: 306
      1 event, rev: 307
    197 progress_notify, rev: 307
      1 event, rev: 308
    352 progress_notify, rev: 308
      1 event, rev: 309
    191 progress_notify, rev: 309
      1 event, rev: 310
      1 progress_notify, rev: 309
    319 progress_notify, rev: 310
      1 event, rev: 311
    183 progress_notify, rev: 311
      1 event, rev: 312
    204 progress_notify, rev: 312
      1 event, rev: 313
    556 progress_notify, rev: 313
      1 event, rev: 314
      1 progress_notify, rev: 313
   1143 progress_notify, rev: 314
      1 event, rev: 315
     70 progress_notify, rev: 315
      1 event, rev: 316
    118 progress_notify, rev: 316
      1 progress_notify, rev: 317
      1 event, rev: 317
    180 progress_notify, rev: 317
      1 event, rev: 318
    377 progress_notify, rev: 318
      1 event, rev: 319
    802 progress_notify, rev: 319
      1 event, rev: 320
     10 progress_notify, rev: 319
    513 progress_notify, rev: 320
      1 event, rev: 321
    189 progress_notify, rev: 321
      1 event, rev: 322
    366 progress_notify, rev: 322
      1 event, rev: 323
      1 progress_notify, rev: 322
    373 progress_notify, rev: 323
      1 event, rev: 324
    195 progress_notify, rev: 324
      1 event, rev: 325
    741 progress_notify, rev: 325
      1 event, rev: 326
    182 progress_notify, rev: 326
      1 event, rev: 327
    209 progress_notify, rev: 327
      1 event, rev: 328
    735 progress_notify, rev: 328
      1 event, rev: 329
      1 progress_notify, rev: 328
    702 progress_notify, rev: 329
      1 event, rev: 330
    149 progress_notify, rev: 330
      2 event, rev: 331
   1125 progress_notify, rev: 331
      1 event, rev: 332
  39861 progress_notify, rev: 332
      1 event, rev: 333
     10 progress_notify, rev: 333
```